plugin/pkg/listener/listener.go

```diff
@@ -0,0 +1,208 @@
+package listener
+
+import (
+	"fmt"
+	"log"
+	"net"
+
+	"github.com/mholt/caddy"
+)
+
+// Booker - generic type providing information on the booker
+type Booker interface {
+	// Tag is used to verify re-usability of a Listener - listener is reusable only for Booker with the same Tag
+	Tag() string
+}
+
+// booking structure use at booking, and allocation time of the listener
+type booking struct {
+	booker    Booker
+	hostname  string
+	port      string
+	protocol  string
+	reusable  *net.Listener
+	allocated *net.Listener
+}
+
+// a way to identify same hostname
```

Comment:
this feels brittle *and* only deals with localhost. Can we get away with just not caring?